Update Modular Pipeline documentation and split Micropackaging into its own page (#1128)
quantumblacklabs/kedro: docs/source/06_nodes_and_pipelines/03_modular_pipelines.md

In many typical Kedro projects, a single (“main”) pipeline increases in complexity as the project evolves. To keep your project fit for purpose, we recommend that you create modular pipelines, which are logically isolated and can be reused. Modular pipelines are easier to develop, test and maintain, and are portable so they can be copied and reused between projects.

> Modular pipelines allow you to instantiate pipelines multiple times, but support overriding inputs/outputs/parameters. They are reusable within the same codebase and shareable across projects via [micro-packaging](04_micro_packaging.md). This is the modern way to use Kedro and will change the way you think about your pipelines.

Reviewer: "Modular pipelines allow you to instantiate pipelines multiple times, but support overriding inputs/outputs/parameters."

Shouldn't it be `and` instead of `but`?

> Modular pipelines allow you to instantiate pipelines multiple times, but support overriding inputs/outputs/parameters. They are reusable within the same codebase and shareable across projects via [micro-packaging](04_micro_packaging.md). This is the modern way to use Kedro and will change the way you think about your pipelines.

Author: ```suggestion
Modular pipelines allow you to instantiate pipelines multiple times whilst allowing the user to override inputs/outputs/parameters. They are reusable within the same codebase and shareable across projects via [micro-packaging](04_micro_packaging.md). This is the modern way to use Kedro and will change the way you think about your pipelines.
```

> Modular pipelines allow you to instantiate pipelines multiple times whilst allowing the user to override inputs/outputs/parameters. They are reusable within the same codebase and shareable across projects via [micro-packaging](04_micro_packaging.md). This is the modern way to use Kedro and will change the way you think about your pipelines.

Author: ```suggestion
Modular pipelines allow you to instantiate pipelines multiple times whilst allowing the user to override inputs/outputs/parameters. They are reusable (with dynamic arguments) within the same codebase and shareable across projects via [micro-packaging](04_micro_packaging.md). This is the modern way to use Kedro and will change the way you think about your pipelines.
```

Author: wdyt?

> Modular pipelines allow you to instantiate pipelines multiple times whilst allowing the user to override inputs/outputs/parameters. They are reusable within the same codebase and shareable across projects via [micro-packaging](04_micro_packaging.md). This is the modern way to use Kedro and will change the way you think about your pipelines.

Author: wdyt?
```suggestion
Modular pipelines allow you to instantiate pipelines multiple times whilst allowing the user to override inputs/outputs/parameters. They are reusable (with dynamic arguments) within the same codebase and shareable across projects via [micro-packaging](04_micro_packaging.md). This is the modern way to use Kedro and will change the way you think about your pipelines.
```

```eval_rst
.. note:: The Kedro project visualised below is representative of one that might be seen in the real world. It takes full advantage of modular pipelines for ``Data Ingestion``, ``Feature Engineering``, ``Reporting`` and ``Train Evaluation`` (which even includes nested instances).
```

### Key concepts

In this section you will learn about how to take advantage of modular pipelines, the key points have been listed below:

1. **A modular pipeline is defined by its folder structure**

   * You can generate this file structure with the CLI command ``kedro pipeline create <pipeline_name>``.
   * The folder structure keeps things isolated and encourages portability.

2. **Modular pipelines are designed to be portable and reusable**

   * It's possible to re-use the same pipeline multiple times within the same project (with different inputs/outputs or parameters).
   * You can also share pipelines across codebases via [micro-packaging](04_micro_packaging.md).

3. **The `kedro.pipeline.modular_pipeline.pipeline` wrapper method unlocks the real power of modular pipelines**

>    * Applying namespaces allows you to simplify your mental model and isolate 'within pipeline' processing steps.

Author: It's a good question, would a reference to the Computer Science-y definition help? 
```suggestion
   * Applying [namespaces](https://en.wikipedia.org/wiki/Namespace) allows you to simplify your mental model and isolate 'within pipeline' processing steps.
```

>    * ``Kedro-Viz`` is able to accelerated development by [rendering namespaced](../03_tutorial/06_visualise_pipeline.md) pipelines as collapsible 'super nodes'.

Reviewer 2: ```suggestion
   * ``Kedro-Viz`` is able to accelerate development by [rendering namespaced](../03_tutorial/06_visualise_pipeline.md) pipelines as collapsible 'super nodes'.
```

<iframe
    src="http://demo.kedro.org"
    width="850",
    height="600"
></iframe>

## How do I create a modular pipeline?

You can use a [project-specific CLI command](../09_development/03_commands_reference.md#kedro-commands) to create a modular pipeline. The pipeline name must adhere to [Python convention](https://realpython.com/python-pep8/#naming-conventions).

```bash
kedro pipeline create <pipeline_name>
```

```eval_rst

> ::note For the full list of available CLI options, you can always run ``kedro pipeline create --help`` for more information.

Reviewer: ```suggestion
.. note:: For the full list of available CLI options, you can always run ``kedro pipeline create --help`` for more information.
```

```

### What does the ``kedro pipeline create`` do?

> Running the pipeline create command adds the following folder structure to your project. For your convenience Kedro gives you a pipeline specific `nodes.py`, `pipeline.py`, parameters and appropriate `tests` structure. You also don't have to add those pesky `__init__.py` files yourself, which is handy 😅.

Reviewer 2: ```suggestion
Running the `pipeline create` command adds the following folder structure to your project. For your convenience Kedro gives you a pipeline-specific `nodes.py`, `pipeline.py`, parameters and appropriate `tests` structure. You also don't have to add those pesky `__init__.py` files yourself, which is handy 😅.
```

<details>
<summary><b>Click to see the generated folder structure</b></summary>

```text
├── conf
│   └── base
│       └── parameters
│           └── {{pipeline_name}}.yml  <-- Pipeline specific parameters
└── src
    ├── my_project
    │   ├── __init__.py
    │   └── pipelines
    │       ├── __init__.py
    │       └── {{pipeline_name}}      <-- This folder defines the modular pipeline
    │           ├── README.md          <-- To store pipeline specific documentation 
    │           ├── __init__.py        <-- So that Python treats this pipeline as a module
    │           ├── nodes.py           <-- To declare your nodes
    │           └── pipeline.py        <-- To structure the pipeline itself
    ├── tests
    │   ├── __init__.py
    │   └── pipelines
    │       ├── __init__.py
    │       └── {{pipeline_name}}      <-- Pipeline specific tests
    │           ├── __init__.py
    │           └── test_pipeline.py

>     └──  pipeline_registry.py          <-- Does NOT automatically update the registry 

Reviewer 2: `pipeline_registry.py` is within `src/my_project`

```

</details>

If you want to do the reverse and remove a modular pipeline, you can use ``kedro pipeline delete`` to do so.

### Ensuring portability

Kedro modular pipelines are shareable between Kedro codebases via [micro-packaging](04_micro_packaging.md), but you need to follow a couple of rules to ensure portability:

* Modular pipelines should **not** depend on the main Python package as this would break portability to another project.

> * Catalog references are not packaged when sharing/consuming modular pipelines.

Author: ```suggestion
* Catalog references are not packaged when sharing/consuming modular pipelines i.e. the `catalog.yml` file is not packaged. As per the [Twelve-Factor App](https://12factor.net/config) philosophy, we package 'business logic' not configuration. 
```

> * Modular pipelines should use [relative imports](https://docs.python.org/3/reference/import.html#package-relative-imports) (as opposed to 'absolute' imports) when accessing their own objects, but are free to use absolute imports otherwise.

Reviewer 2: I didn't think that was still true since using `rope`? Unless I'm misreading the sentence.

Author: I just took it from the old stuff so will remove.
```suggestion
```

* Kedro will only look for top level configuration `/conf/`, therefore providing a configuration folder within the pipeline folder will have no effect.

> * It is recommended that you document the configuration required (parameters and catalog) in the local `README.md` file for any downstream consumers.

Author: I think this is lifted from existing docs - but I think as part of a handover docs this makes sense

> * Kedro will only look for top level configuration `/conf/` providing a configuration folder within the pipeline folder will have no effect.

Reviewer 2: ```suggestion
* Kedro will only look for top level configuration `/conf/`, therefore providing a configuration folder within the pipeline folder will have no effect.
```

### Providing modular pipeline specific dependencies

* A modular pipeline **may** have external dependencies specified in a local `requirements.txt` file.

> * Pipeline specific dependencies are scooped up during [micro-packaging](04_micro_packaging.md) process.

Reviewer: ```suggestion
* Pipeline specific dependencies are scooped up during the [micro-packaging](04_micro_packaging.md) process.
```

* These dependencies are _not_ currently installed by the [`kedro install`](../09_development/03_commands_reference.md#install-all-package-dependencies) command and will have to be manually installed.

## Using the modular `pipeline()` wrapper to provide overrides

This wrapper really unlocks the power of modular pipelines.

* It allows you to start instantiating the same pipeline multiple times.
* These will be static in terms of structure, but dynamic in terms of `inputs`/`outputs`/`parameters`.

> * It also allows you to simplify both your mental models and Kedro-Viz visualisations via `namespace`s.

Reviewer: ```suggestion
* It also allows you to simplify both your mental models and Kedro-Viz visualisations via `namespaces`.
```

Author: @idanov I think this is covered later in more detail

+--------------------+---------------------------------------------------------------------+
| ``namespace``      | The namespace that will be encapsulated by this pipeline instance   |
+--------------------+---------------------------------------------------------------------+
```

## Combining disconnected pipelines

Sometimes two pipelines need to be connected, but do not share any catalog dependencies. The wrapper can be used to solve that.

<details>
<summary>Click here to see a worked example</summary>

```python
cook_pipeline = Pipeline(
    [
        node(
            func=defrost, 
            inputs="frozen_veg", 
            outputs="veg"
        ), 
        node(
            func=grill, 
            inputs="veg", 
            outputs="grilled_veg"
        )
    ]
)

lunch_pipeline = Pipeline(
    [
        node(
            func=eat, 
            inputs="food", 
            outputs=None
        )
    ]
)

cook_pipeline + lunch_pipeline 
```

This combination will visualise since it's valid pre-runtime, but it will not run since `food` is not an output of the cook pipeline:

![disjoined](../meta/images/cook_disjointed.png)

> * Combining `cook_pipeline + lunch_pipeline` will not work since because the `food` doesn't exist as an output of the `cook_pipeline`.

Reviewer 2: ```suggestion
* Combining `cook_pipeline + lunch_pipeline` will not work since `food` doesn't exist as an output of the `cook_pipeline`.
```

Can we actually remove this line, since it's repeated on L168?

* In this case, we will need to map `grilled_veg` to the expected input of `food`.

The wrapper allows us to provide a mapping and fix this disconnect.

```python
from kedro.pipeline.modular_pipeline import pipeline

prep_pipeline = pipeline(
        pipe=cook_pipeline, 
        inputs={"food" : "grilled_veg"} 

>     )

Reviewer 2: ```suggestion
)
```

Reviewer 2: It might be even short enough to leave on one line actually.

meal_pipeline = prep_pipeline + lunch_pipeline
```

> Providing  this input/output override will join up the pipeline nicely:

Reviewer 2: ```suggestion
Providing this input/output override will join up the pipeline nicely:
```

> ![joineted](../meta/images/cook_joined.png)

Reviewer 2: ```suggestion
![joined](../meta/images/cook_joined.png)
```

```eval_rst
.. note:: In this example we have used the ``+`` operator to join two pipelines. Remember you can also use ``sum()`` or pass a list of pipelines to the ``Pipeline()`` constructor as well.
```

</details>

## Using a modular pipeline multiple times

> Reusing pipelines for slightly different purposes can be a real accelerator for teams and organisations when they reach a certain scale.

Author: Beyond this 
> Reusing pipelines for slightly different purposes can be a real accelerator for teams and organisations when they reach a certain scale.

or do you mean in the sub-title?

> Reusing pipelines for slightly different purposes can be real accelerator for teams and organisations when they reach a certain scale.

Reviewer: ```suggestion
Reusing pipelines for slightly different purposes can be a real accelerator for teams and organisations when they reach a certain scale.
```

* Namespaces allow you to '[instantiate](https://en.wikipedia.org/wiki/Instance_(computer_science))' the same pipeline multiple times and keep operations isolated.
* Like one provides arguments to a class' constructor, you can provide overriding inputs/outputs/parameters to the `pipeline()` wrapper

<details>

> <summary>Click here to see a worked example</summary>

Author: I've had a go - wdyt?

<details>
<summary>Click here to see a worked example</summary>

```python
cook_pipeline = Pipeline(
    [
        node(
            func=defrost, 
            inputs="frozen_veg", 
            outputs="veg", 
            name="defrost_node"
        ),
        node(
            func=grill, 
            inputs="veg", 
            outputs="grilled_veg"
        ),
    ]
)

eat_breakfast_pipeline = Pipeline(
    [
        node(
            func=eat_breakfast, 
            inputs="breakfast_food", 
            outputs=None
        )
    ]
)
eat_lunch_pipeline = Pipeline(
    [
        node(
            func=eat_lunch, 
            inputs="lunch_food", 
            outputs=None
        )
    ]
)

cook_pipeline + eat_breakfast_pipeline + eat_lunch_pipeline 
```

If we visualise the snippet above, we see a disjointed pipeline:

* We need to "defrost" two different types of food via different pipelines.
* We can't use the `cook_pipeline` twice because the internal dataset names will conflict.
* Mapping all datasets via the `pipeline()`  wrapper will also cause conflicts.

> ![cook no namespace](../meta/images/cook_no_namespace.png)

Reviewer 3: I just realised this image has "Frozen Meat" on it and not "Vegetables" like in the example.

Author: Good spot! I made everything vegan when I was writing this :joy:

Author: Fixed!

<details>
<summary>Click here to see a worked example</summary>

* We instantiate the `template_pipeline` twice, but pass in different parameters.
* `input1` and `input2` are 'frozen' and thus shared in both instances.
* `params:override_me` does not actually exist and is designed to be overridden in both cases.
* Providing a namespace isolates the intermediate operation and visualises nicely.

```python
template_pipeline = Pipeline(
    [
        node(
            func=node_func1, 
            inputs=["input1", "input2", "params:override_me"], 
            outputs="intermediary_output"
        ),
        node(
            func=node_func2, 
            inputs="intermediary_output", 
            outputs="output"
        ),
    ]
)

alpha_pipeline = pipeline(
    pipe=template_pipeline,
    inputs={"input1", "input2"},
    parameters={"params:override_me": "params:alpha"}
    namespace="alpha"
)

beta_pipeline = pipeline(
    pipe=template_pipeline,
    inputs={"input1", "input2"},
    parameters={"params:override_me": "params:beta"},
    namespace="beta",
)

final_pipeline = alpha_pipeline + beta_pipeline
```

</details>

> ![namespaced_params](../meta/images/cook_params.png)

Reviewer 2: Can we include a visualisation that also shows that the different parameters are used in the two pipelines? This one only shows two modular pipelines that take the same inputs `input1` and `input2`.
